File: backend/app/schemas/planning.py
"""
Planning module schemas (Pydantic) for PRIA v7
Request and response validation
"""
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime, date


class MomentoRequest(BaseModel):
    """Request to create/update a Momento."""
    nombre: str = Field(..., description="Momento name: Inicio, Desarrollo, or Cierre")
    duration_minutes: int = Field(default=15, description="Duration in minutes")
    content_text: Optional[str] = Field(None, description="Main lesson content")
    recursos: List[str] = Field(default=[], description="List of resources needed")
    evaluacion: Optional[str] = Field(None, description="Assessment for this moment")


class MomentoResponse(BaseModel):
    """Response with Momento details."""
    id: int
    order: int
    nombre: str
    duration_minutes: int
    content_text: Optional[str]
    recursos: List[str]
    evaluacion: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MicroObjetivoRequest(BaseModel):
    """Request to create/update a MicroObjetivo."""
    texto: str = Field(..., description="SMART objective text")
    verificable: bool = Field(default=True, description="Is it measurable?")
    prioridad: str = Field(default="normal", description="baja, normal, or alta")
    depende_de: Optional[int] = Field(None, description="ID of parent objective")


class MicroObjetivoResponse(BaseModel):
    """Response with MicroObjetivo details."""
    id: int
    texto: str
    verificable: bool
    completado: bool
    prioridad: str
    depende_de: Optional[int]
    created_at: datetime

    class Config:
        from_attributes = True


class CreateWeekRequest(BaseModel):
    """Request to create a weekly plan."""
    pdc_id: int
    week_number: int = Field(..., description="Week number 15-30")
    subject: Optional[str] = None
    grade_level: Optional[str] = None
    status: str = Field(default="draft", description="draft, published, or completed")


class UpdateWeekRequest(BaseModel):
    """Request to update a weekly plan."""
    subject: Optional[str] = None
    grade_level: Optional[str] = None
    status: Optional[str] = None


class WeeklyPlanResponse(BaseModel):
    """Response with full weekly plan including momentos."""
    id: int
    pdc_id: int
    week_number: int
    subject: Optional[str]
    grade_level: Optional[str]
    status: str
    momentos: List[MomentoResponse] = []
    micro_objetivos: List[MicroObjetivoResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GenerateWeekRequest(BaseModel):
    """Request to auto-generate a weekly plan."""
    pdc_id: int
    week_number: Optional[int] = None  # If None, generate all 16 weeks
    profile_overrides: Optional[dict] = None


class GenerationJobResponse(BaseModel):
    """Response for async generation job."""
    job_id: str = Field(..., description="Celery task ID")
    status: str = Field(default="queued", description="queued, processing, complete, failed")
    progress: int = Field(default=0, description="0-100")
    weeks_completed: int = Field(default=0)
    weeks_total: int = Field(default=1)
    eta: Optional[float] = None


class CalendarEventResponse(BaseModel):
    """Response with calendar event details."""
    id: int
    fecha: date
    nombre_evento: str
    tipo: str  # vacation, holiday, event
    descripcion: Optional[str]

    class Config:
        from_attributes = True


class CopyWeekRequest(BaseModel):
    """Request to copy a week to another week."""
    target_week_id: int = Field(..., description="Target weekly plan ID")

File: frontend/app/lib/types/planning.ts
/**
 * Planning Module TypeScript Types
 * Defines all interfaces for weekly planning and moment-based instruction
 */

// Momento - Instructional phase (Inicio, Desarrollo, Cierre)
export interface Momento {
  id: string;
  week_id: string;
  order: 1 | 2 | 3; // Position in sequence
  nombre: 'Inicio' | 'Desarrollo' | 'Cierre'; // Phase name
  duration_minutes: number; // Instructional time
  content_text: string; // Main content (accepts markdown/rich text)
  recursos: string[]; // List of required resources/materials
  evaluacion: string; // Assessment/evaluation strategy
  created_at?: string;
  updated_at?: string;
}

// Micro-objetivo - Learning objective that can be verified
export interface MicroObjetivo {
  id: string;
  week_id: string;
  texto: string; // Objective description
  verificable: string; // How to verify achievement
  completado: boolean; // Completion status
  prioridad: 'baja' | 'normal' | 'alta'; // Priority level
  depende_de: string[]; // IDs of objectives this depends on
  created_at?: string;
  updated_at?: string;
}

// Week - Weekly planning unit
export interface Week {
  id: string;
  pdc_id: string;
  number: number; // Week 15-30 (or custom range)
  subject: string; // Subject name
  grade_level: string; // Grade/level
  status: 'draft' | 'published' | 'completed'; // Publication state
  momentos: Momento[]; // Array of 3 momentos (Inicio, Desarrollo, Cierre)
  micro_objetivos?: MicroObjetivo[]; // Associated learning objectives
  created_at: string;
  updated_at: string;
}

// Calendar event - School calendar item
export interface CalendarEvent {
  date: string; // ISO 8601 date
  week_number: number; // Associated week (nullable for holidays)
  event_name: string; // Event title
  event_type: 'vacation' | 'holiday' | 'event'; // Event classification
}

// Request to generate weekly plans
export interface GenerateWeekRequest {
  pdc_id: string;
  week_number: number;
  profile_overrides?: {
    grade_level?: string;
    // Other profile customizations
  };
}

// Job status tracking for async generation
export interface GenerationJob {
  job_id: string;
  status: 'queued' | 'processing' | 'completed' | 'failed';
  progress: number; // 0-100
  current_week?: number; // Currently processing week
  total_weeks: number;
  error_message?: string;
  started_at: string;
  completed_at?: string;
}

// Request to start batch generation
export interface StartGenerationRequest {
  pdc_id: string;
  profile_overrides?: {
    grade_level?: string;
  };
}

// API Response wrapper
export interface ApiResponse<T> {
  data: T;
  success: boolean;
  error?: string;
}

// List response with items
export interface ListResponse<T> {
  items: T[];
  total: number;
  skip: number;
  limit: number;
}

// Update Week request
export interface UpdateWeekRequest {
  subject?: string;
  grade_level?: string;
  status?: 'draft' | 'published' | 'completed';
}

// Create/Update Momento request
export interface CreateMomentoRequest {
  nombre: 'Inicio' | 'Desarrollo' | 'Cierre';
  duration_minutes: number;
  content_text: string;
  recursos: string[];
  evaluacion: string;
}

export interface UpdateMomentoRequest extends Partial<CreateMomentoRequest> {}

// Copy week request
export interface CopyWeekRequest {
  source_week_id: string;
  target_week_id: string;
}
